### PR TITLE
Introduce OAuth login flow

### LINKED ISSUE #553: Move request for email confirmation into webview
Depends on #242 (OAuth)

- [x] Ensure that the email confirmation step is tied to the authentication webview rather than on the entries list empty state, so the app doesn't give false expectations around being "Finished" with setup

- [ ] Consider removing the email verification from the empty state entries list (unless we want it as a fallback in instances where someone dismisses the webview and still hasn't confirmed email)

### LINKED ISSUE #555: FxA language confusing for Lockbox users
Some of the language around FxA verifcation is confusing, because it is not Lockbox specific.

Depends on #242 

- [x] Change "Continue to Firefox Sync" in **sign-in webview** to "Continue to Firefox Lockbox"
- [x] Change "Confirm Sign in to Firefox" in **email** to "Confirm Sign-in to Firefox Lockbox" 

_This may need to be tied to FxA efforts if it needs to be handled on their end, but placing within our repo for now so we can track against it._

### LINKED ISSUE #556: Switch to decoupled email FxA flow
Depends on #242 (OAuth) and https://github.com/mozilla/application-services/issues/124

Some UR participants did not immediately know if they had an FxA account, when authenticating with Lockbox for the first time.

_Pending the completion of OAuth..._

- [x] Switch to decoupled, email-first FxA flow, instead of standard "sign-in" with email and password

### PR DESCRIPTION
**Pull request by jhugman**
This PR uses the rust client of https://github.com/mozilla/application-services . 

It uses that client to obtain the URL for login, the key material for syncing (access token, scoped keys), and user profile (avatar, email address etc).

These were previously obtained from `lockbox-ios-fxa-sync`. That library has been altered in https://github.com/mozilla-lockbox/lockbox-ios-fxa-sync/pull/10 to support this new login flow.

This PR should not be merged until that PR is merged, as `Cartfile` and `Cartfile.resolved` need to be updated to point to the merged branch.

Some work is still left; much of the work on #412 needs to be at least re-verified, if not redone.

Fixes https://github.com/mozilla-lockbox/lockbox-ios/issues/242
Fixes #553 
Fixes #555
Fixes #556

### CODE REVIEW

**Comment by eoger** [comment]
Just a quick note to let you know that the version of fxa-client (0.1.0) used in that PR does synchronous network calls. I'm going make a new release of the SDK in the following days with async methods. Let me know if you want to sync up on Slack.

**Comment by eoger** [comment]
Version 0.2.0 of the FxA SDK has been released

File: lockbox-iosTests/AccountSettingPresenterSpec.swift [hunk 1]
```diff
@@ -105,65 +106,11 @@ class AccountSettingPresenterSpec: QuickSpec {
...
-                                            .email(email)
-                                            .build()
-                            )
-                        }
-
-                        it("tells the view to display the email") {
-                            expect(self.view.displayNameObserver.events.last!.value.element).to(equal(email))
-                        }
-                    }
+                    // tricky to test because we can't construct FxAClient.Profile
```
Comment (linuxwolf) [inline lockbox-iosTests/AccountSettingPresenterSpec.swift]: Maybe file an enhancement request with App Services?

File: lockbox-iosTests/AccountStoreSpec.swift [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                            .bind(to: urlObserver)
+                            .disposed(by: self.disposeBag)
+                }
+
+                it("populates the loginURL for the Lockbox configuration on initialization") {
+                    if let config = try? FxAConfig.custom(content_base: "https://accounts.firefox.com"),
+                       let fxa = try? FirefoxAccount(config: config, clientId: Constant.fxa.clientID),
+                       let url = try? fxa.beginOAuthFlow(
+                               redirectURI: Constant.fxa.redirectURI,
+                               scopes: [
```
Comment (linuxwolf) [inline lockbox-iosTests/AccountStoreSpec.swift]: see about using `Constants.swift` for this.

File: lockbox-iosTests/AccountStoreSpec.swift [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
...s.firefox.com/v1/jwks\",\"token_endpoint\":\"https://oauth.accounts.firefox.com/v1/token\",\"userinfo_endpoint\":\"https://profile.accounts.firefox.com/v1/profile\"},\"login_state\":\"Unknown\",\"oauth_cache\":{\"profile https://identity.mozilla.com/apps/oldsync https://identity.mozilla.com/apps/lockbox\":{\"access_token\":\"42e4bacc8affd23a4ae264fd59ae8ec4a58d5af03b1215daa9cd940f6f80bd9e\",\"keys\":\"{\\\"https://identity.mozilla.com/apps/oldsync\\\":{\\\"kty\\\":\\\"oct\\\",\\\"scope\\\":\\\"https://identity.mozilla.com/apps/oldsync\\\",\\\"k\\\":\\\"VEZDYJ3Jd1Ui0ZVtW8pPHD6LZ48Jd30p-y-PLQQYa0PRcMZtiM6zJO4_I2lxEg__qkxXldPyLiM5PYY9VBD64w\\\",\\\"kid\\\":\\\"1519160140602-WMF1HOhJbtMVueuy3tV4vA\\\"},\\\"https://identity.mozilla.com/apps/lockbox\\\":{\\\"kty\\\":\\\"oct\\\",\\\"scope\\\":\\\"https://identity.mozilla.com/apps/lockbox\\\",\\\"k\\\":\\\"oGGfsZk8xMXtBzGzy2WY3QGPNOTer0VGIC3Uyz9Jy9w\\\",\\\"kid\\\":\\\"1519160141-YqmShzWPQhHp0RNiZs25zg\\\"}}\",\"refresh_token\":\"fd417de6f90683a046070450ead9844a095bee5df934cc0fe3455bd2d41d394a\",\"expires_at\":1531864458,\"scopes\":[\"profile\",\"https://identity.mozilla.com/apps/oldsync\",\"https://identity.mozilla.com/apps/lockbox\"]}}}"
```
Comment (linuxwolf) [inline lockbox-iosTests/AccountStoreSpec.swift]: Using real results is helpful, but we should confirm the account this real results maps to doesn't hold real data for real people.

File: lockbox-iosTests/AccountStoreSpec.swift [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
```
Comment (linuxwolf) [inline lockbox-iosTests/AccountStoreSpec.swift]: same as above.

File: lockbox-iosTests/AccountStoreSpec.swift [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                                dispatcher: self.dispatcher,
+                                keychainWrapper: self.keychainManager
+                        )
+
+                        self.subject.oauthInfo
+                                .bind(to: oauthObserver)
+                                .disposed(by: self.disposeBag)
+                    }
+
+                    it("pushes a non-nil profile") {
```
Comment (linuxwolf) [inline lockbox-iosTests/AccountStoreSpec.swift]: nit: s/profile/oauthInfo/

File: lockbox-iosTests/AccountStoreSpec.swift [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                }
+
+                describe("when the keychain does not have a valid fxa account") {
+                    beforeEach {
+                        self.subject.oauthInfo
+                                .bind(to: oauthObserver)
+                                .disposed(by: self.disposeBag)
+                    }
+
+                    it("pushes a nil profile") {
```
Comment (linuxwolf) [inline lockbox-iosTests/AccountStoreSpec.swift]: nit: s/profile/oauthInfo/

File: lockbox-ios/Action/DataStoreAction.swift [hunk 7]
```diff
@@ -23,8 +24,7 @@ enum DataStoreAction: Action {
 extension DataStoreAction: Equatable {
     static func ==(lhs: DataStoreAction, rhs: DataStoreAction) -> Bool {
         switch (lhs, rhs) {
-        case (.initialize(let lhBlob), .initialize(let rhBlob)):
-            return lhBlob == rhBlob
+        case (.updateCredentials, .updateCredentials): return true // TODO equality
```
[the remark this answers is not in the document]
Comment (sashei) [reply lockbox-ios/Action/DataStoreAction.swift]: There's a larger question here about the extensibility of the `fxa-client` library; I've added this to the agenda for our shared meeting.

File: lockbox-ios/Store/AccountStore.swift [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+        self.generateInitialURL()
+        self.populateInitialValues()
+    }
+}
+
+extension AccountStore {
+    private func generateInitialURL() {
+        // TODO: it doesn't look like this library allows us to specify "signin" action?
+        // the page is defaulting to signup :(
```
[the remark this answers is not in the document]
Comment (sashei) [reply lockbox-ios/Store/AccountStore.swift]: Follow-up issue: mozilla/application-services#124 and #569

File: lockbox-iosTests/AccountSettingPresenterSpec.swift (hunk 1 above)
Comment (sashei) [reply lockbox-iosTests/AccountSettingPresenterSpec.swift]: follow-up question in the meeting agenda for tomorrow

**Comment by devinreams** [comment]
Noted during a quick backlog review: PR is still not done, @sashei still working to get the bump up to `0.2.0` working.

File: Cartfile [hunk 9]
```diff
@@ -4,6 +4,7 @@ github "ReactiveX/RxSwift"
 github "RxSwiftCommunity/RxDataSources" ~> 3.0
 github "RxSwiftCommunity/RxOptional" ~> 3.1.3
 github "mozilla-mobile/telemetry-ios"
-github "mozilla-lockbox/lockbox-ios-fxa-sync"
 github "jrendel/SwiftKeychainWrapper" ~> 3.0
-github "mozilla-mobile/MappaMundi" "master"
\ No newline at end of file
+github "mozilla-mobile/MappaMundi" "master"
+github "mozilla/application-services" "0.2.0"
+github "mozilla-lockbox/lockbox-ios-fxa-sync" "044d84bfbe44e6fb43e4b7737cf551ca77a918da"
```
Comment (devinreams) [inline Cartfile]: Reminder to merge https://github.com/mozilla-lockbox/lockbox-ios-fxa-sync/pull/10 and change this back to no SHA or at the latest SHA, whichever is preferred.

File: lockbox-iosTests/WelcomePresenterSpec.swift [hunk 10]
```diff
@@ -315,7 +316,7 @@ class WelcomePresenterSpec: QuickSpec {
                                 }
 
                                 it("starts authentication") {
-                                    expect(self.biometryManager.authMessage).to(equal(email))
+//                                    expect(self.biometryManager.authMessage).to(equal(email))
```
Comment (linuxwolf) [inline lockbox-iosTests/WelcomePresenterSpec.swift]: Is this intentional, or a holdover from troubleshooting?

Comment (sashei) [reply lockbox-iosTests/WelcomePresenterSpec.swift]: holdover! good catch, thanks @linuxwolf

**Comment by sashei** [comment]
One thing to note is that all the integration specs are failing, haven't looked into this but I assume that it's because authentication isn't working at all / no longer looks like what the specs expect. If @jhugman wants to pick that up and fix them that would be great; otherwise I'll try to get to them tomorrow.

**Comment by devinreams** [comment]
Would it help if I did some end-to-end functional testing on this branch before we merge? If so please tag me as a reviewer again and I'll get the `lockbox-ios-fxa-sync` library updated locally and run through some tests...

**Comment by sashei** [comment]
@devinreams yes; let me get this branch rebased from the #527 merge and then I will tag you.

File: lockbox-ios/Store/AccountStore.swift [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+              let state = dic["state"] else {
+            return
+        }
+
+        self.fxa?.completeOAuthFlow(code: code, state: state) { (info: OAuthInfo?, _) in
+            self._oauthInfo.onNext(info)
+
+            if let accountJSON = try? self.fxa?.toJSON(), accountJSON != nil {
+                self.keychainWrapper.set(accountJSON!, forKey: KeychainKey.accountJSON.rawValue)
+            }
```
Comment (jhugman) [inline lockbox-ios/Store/AccountStore.swift]: Nit: everytime `!` is used in swift, an angel loses its wings.

Also: `??` should collapse under its own uncertainty.

```swift
            if let opt = try? self.fxa?.toJSON(), let accountJSON = opt {
                self.keychainWrapper.set(accountJSON, forKey: KeychainKey.accountJSON.rawValue)
            }
```

File: lockbox-ios/Store/AutoLockStore.swift [hunk 12]
```diff
@@ -74,6 +74,8 @@ extension AutoLockStore {
                 .take(1)
                 .subscribe(onNext: { (latest: AutoLockSetting) in
                     switch latest {
+                    case .OneMinute:
+                        self.setTimer(seconds: 60)
```
Comment (jhugman) [inline lockbox-ios/Store/AutoLockStore.swift]: 👏

File: lockbox-ios/Store/DataStore.swift [hunk 13]
```diff
@@ -92,7 +95,8 @@ class DataStore {
     private let fxaLoginHelper: FxALoginHelper
     private let profileFactory: ProfileFactory
     private let keychainWrapper: KeychainWrapper
-    private var profile: Profile
+    private var profile: FxAUtils.Profile
```
[the remark this answers is not in the document]
Comment (jhugman) [reply lockbox-ios/Store/DataStore.swift]: `FxAClient.Profile` is the struct for user account profile details: display name, avatar url, email address.

`FxAUtils.Profile` is akin to the browser profile directory… an artifact of the provenance of `lockbox-ios-fxa-sync`.

Agreed, it should have a comment, but perhaps should also stop referring to `FxAUtils.Profile` as profile.

File: lockbox-ios/Store/DataStore.swift [hunk 14]
```diff
@@ -389,9 +399,8 @@ extension DataStore {
             if !profile.hasAccount() {
                 // first run.
                 self.storageStateSubject.onNext(.Unprepared)
-            } else {
-                self.storageStateSubject.onNext(.Preparing)
             }
+
             self.syncSubject.onNext(.NotSyncable)
```
Comment (jhugman) [inline lockbox-ios/Store/DataStore.swift]: How do we test this state machine? This should probably be the subject of a followup PR.

Comment (sashei) [reply lockbox-ios/Store/DataStore.swift]: `DataStore.swift` is not tested at all right now :p

There is an issue that sort of addresses this that we'll talk about in tech debt triage next week, but maybe we need another one.

**Comment by nickbrandt** [comment]
@devinreams @jhugman @sashei This is all looking really good! I ran into a couple issues though that I wanted to point out:

I disconnected my account after signing in, and when I went back to go through the Get Started flow again, it brought me here (and appears I'm not signed out):
account


After clicking "Sign Out" from there, it took me to the FxA sign-in (not email first flow). I entered my credentials, and upon clicking "Finish" on the "Welcome to Lockbox" screen, it brought me here (looks like the No Matching Results screen for when searching):
no-results


The only other small thing I noticed was when going through the FxA email first flow, even though the fields appeared focused, I had to click into the fields to bring up the keyboard. This might be a limitation that we have, but just wanted to verify.

**Comment by devinreams** [comment]
>After clicking "Sign Out" from there, it took me to the FxA sign-in (not email first flow). I entered my credentials, and upon clicking "Finish" on the "Welcome to Lockbox" screen, it brought me here (looks like the No Matching Results screen for when searching):

**Comment by changecourse** [comment]
@devinreams #422 addresses the account still being active... wondering about this other issue raised by Nick in screenshot 2, which seemingly displays both the wrong screen (search results, no entries found) and also may prevent users from being able to display their entries again (as I don't think that an email confirmation was holding that up)

**Comment by devinreams** [comment]
>which seemingly displays both the wrong screen (search results, no entries found) and also may prevent users from being able to display their entries again

Yup, understood. This problem was introduced and coincided with me bumping up to the `0.2.1` version that also introduced the email-first flow, so I believe that's related to the manual "sign out" from the web view.

I think we want to get the proper timing of cookie/data clearing  "Disconnect Now" in place and figured out first and _then_ go through and make sure no other unexpected data was cleared resulting in that weird "no entries" state.

My thinking is: since the manual "sign out" from the FxA website itself is not expected, but the introduction of it is likely causing the data issue, let's not try to solve for it until we're sure that is (or is not) the cause of the problem... cool?

**Comment by changecourse** [comment]
Yep, makes total sense, thanks for the clarification!

**Pull request merged**